### PR TITLE
Add new Account SAS Token cmdlet, and add IPacl/Protocal support to B/T/Q/F new SAS cmdlets

### PR DESCRIPTION
**Pull request by blueww**
- Add cmdlet to generate SAS token against storage account
  - New-AzureStorageAccountSASToken
- Add IPAddressOrRange/Protocol support in cmdlets to generate SAS token against blob, container, file, share, table, queue
  - New-AzureStorageBlobSASToken
  - New-AzureStorageContainerSASToken
  - New-AzureStorageFileSASToken
  - New-AzureStorageShareSASToken
  - New-AzureStorageQueueSASToken
  - New-AzureStorageTableSASToken
    *Change the Help xml file for the cmdlets update. (The help text is pending Tamra review as he is OOF, might have another PR for Help Text change from Tamra later)
    *The Help xml has many changes as it's re-generated from the PSHelp Edit Tool. You can find all Help text change (and conclusion of each change)  from: 
    [HelpChange.docx](https://github.com/Azure/azure-powershell/files/135817/HelpChange.docx).

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@blueww__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <p>
        It looks like you're working at Microsoft (weiwei). If you're full-time, we DON'T require a contribution license agreement.
    </p>
    <p>
        If you are a vendor, DO please sign the electronic contribution license agreement. It will take 2 minutes and there's no faxing! https://cla.azure.com.
    </p>

TTYL, AZPRBOT;

File: src/Common/Storage/Commands.Storage/Blob/Cmdlet/NewAzureStorageContainerSasToken.cs [hunk 1]
```diff
@@ -54,6 +55,12 @@ public string Policy
             ParameterSetName = SasPermissionParameterSet)]
```
Comment (markcowl) [inline src/Common/Storage/Commands.Storage/Blob/Cmdlet/NewAzureStorageContainerSasToken.cs]: indentation

File: src/Common/Storage/Commands.Storage/Blob/Cmdlet/NewAzureStorageBlobSasToken.cs [hunk 2]
```diff
@@ -80,6 +81,12 @@ public string Policy
             ParameterSetName = BlobPipelineParameterSetWithPermision)]
         public string Permission { get; set; }
 
+        [Parameter(Mandatory = false, HelpMessage = "Protocol can be used in the request with this SAS token.")]
+        public SharedAccessProtocol Protocol { get; set; }
```
Comment (markcowl) [inline src/Common/Storage/Commands.Storage/Blob/Cmdlet/NewAzureStorageBlobSasToken.cs]: This is not an appropriate parameter type - as a user, it is unclear to me how to create this type.  This needs to be a simple type, or a type that can be easily created from another cmdlet.  Preferrably a simple type.

File: src/Common/Storage/Commands.Storage/Blob/Cmdlet/NewAzureStorageContainerSasToken.cs [hunk 3]
```diff
@@ -54,6 +55,12 @@ public string Policy
...
         public string Permission { get; set; }
 
+        [Parameter(Mandatory = false, HelpMessage = "Protocol can be used in the request with this SAS token.")]
+        public SharedAccessProtocol Protocol { get; set; }
```
Comment (markcowl) [inline src/Common/Storage/Commands.Storage/Blob/Cmdlet/NewAzureStorageContainerSasToken.cs]: same comment.  How do I create a SharedAccessProtocol

File: src/Common/Storage/Commands.Storage/Common/Cmdlet/NewAzureStorageAccountSasToken.cs [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+        public SharedAccessAccountServices Service { get; set; }
+
+        [Parameter(Mandatory = true, HelpMessage = "Resource type that this SAS token applies to.")]
+        public SharedAccessAccountResourceTypes ResourceType { get; set; }
+
+        [Parameter(Mandatory = true, HelpMessage = "Permissions.")]
+        public string Permission { get; set; }
+
+        [Parameter(Mandatory = false, HelpMessage = "Protocol can be used in the request with this SAS token.")]
+        public SharedAccessProtocol Protocol { get; set; }
```
Comment (markcowl) [inline src/Common/Storage/Commands.Storage/Common/Cmdlet/NewAzureStorageAccountSasToken.cs]: same comment

File: src/Common/Storage/Commands.Storage/Common/Cmdlet/NewAzureStorageAccountSasToken.cs [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+    using System.Management.Automation;
+    using System.Security.Permissions;
+    using Microsoft.WindowsAzure.Commands.Storage.Model.Contract;
+    using Microsoft.WindowsAzure.Storage;
+
+    [Cmdlet(VerbsCommon.New, StorageNouns.AccountSas), OutputType(typeof(String))]
+    public class NewAzureStorageAccountSasTokenCommand : StorageCloudBlobCmdletBase
+    {
+        [Parameter(Mandatory = true, HelpMessage = "Service type that this SAS token applies to.")]
+        public SharedAccessAccountServices Service { get; set; }
```
Comment (markcowl) [inline src/Common/Storage/Commands.Storage/Common/Cmdlet/NewAzureStorageAccountSasToken.cs]: How do I create this type?

File: src/Common/Storage/Commands.Storage/Common/Cmdlet/NewAzureStorageAccountSasToken.cs [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+
+        [Parameter(Mandatory = true, HelpMessage = "Resource type that this SAS token applies to.")]
+        public SharedAccessAccountResourceTypes ResourceType { get; set; }
```
Comment (markcowl) [inline src/Common/Storage/Commands.Storage/Common/Cmdlet/NewAzureStorageAccountSasToken.cs]: How do I create this type?

File: src/Common/Storage/Commands.Storage/Common/Cmdlet/NewAzureStorageAccountSasToken.cs [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+        public SharedAccessProtocol Protocol { get; set; }
+
+        [Parameter(Mandatory = false, HelpMessage = "IP, or IP range ACL (access control list) that the request would be accepted from by Azure Storage.")]
+        public string IPAddressOrRange { get; set; }
+
+        [Parameter(Mandatory = false, HelpMessage = "Start Time")]
+        public DateTime? StartTime { get; set; }
+
+        [Parameter(Mandatory = false, HelpMessage = "Expiry Time")]
+        public DateTime? ExpiryTime { get; set; }
```
Comment (markcowl) [inline src/Common/Storage/Commands.Storage/Common/Cmdlet/NewAzureStorageAccountSasToken.cs]: ExpirationTime

File: src/Common/Storage/Commands.Storage/File/Cmdlet/NewAzureStorageFileSasToken.cs [hunk 8]
```diff
@@ -104,6 +105,12 @@ public string Policy
             ParameterSetName = CloudFileSasPermissionParameterSet)]
         public string Permission { get; set; }
 
+        [Parameter(Mandatory = false, HelpMessage = "Protocol can be used in the request with this SAS token.")]
+        public SharedAccessProtocol Protocol { get; set; }
```
Comment (markcowl) [inline src/Common/Storage/Commands.Storage/File/Cmdlet/NewAzureStorageFileSasToken.cs]: same comment

File: src/Common/Storage/Commands.Storage/File/Cmdlet/NewAzureStorageShareSasToken.cs [hunk 9]
```diff
@@ -56,6 +57,12 @@ public string Policy
             ParameterSetName = SasPermissionParameterSet)]
         public string Permission { get; set; }
 
+        [Parameter(Mandatory = false, HelpMessage = "Protocol can be used in the request with this SAS token.")]
+        public SharedAccessProtocol Protocol { get; set; }
```
Comment (markcowl) [inline src/Common/Storage/Commands.Storage/File/Cmdlet/NewAzureStorageShareSasToken.cs]: same

File: src/Common/Storage/Commands.Storage/Queue/Cmdlet/NewAzureStorageQueueSasToken.cs [hunk 10]
```diff
@@ -54,6 +55,12 @@ public string Policy
             ParameterSetName = SasPermissionParameterSet)]
         public string Permission { get; set; }
 
+        [Parameter(Mandatory = false, HelpMessage = "Protocol can be used in the request with this SAS token.")]
+        public SharedAccessProtocol Protocol { get; set; }
```
Comment (markcowl) [inline src/Common/Storage/Commands.Storage/Queue/Cmdlet/NewAzureStorageQueueSasToken.cs]: same comment

File: src/Common/Storage/Commands.Storage/Table/Cmdlet/NewAzureStorageTableSasToken.cs [hunk 11]
```diff
@@ -55,6 +56,12 @@ public string Policy
             ParameterSetName = SasPermissionParameterSet)]
         public string Permission { get; set; }
 
+        [Parameter(Mandatory = false, HelpMessage = "Protocol can be used in the request with this SAS token.")]
+        public SharedAccessProtocol Protocol { get; set; }
```
Comment (markcowl) [inline src/Common/Storage/Commands.Storage/Table/Cmdlet/NewAzureStorageTableSasToken.cs]: same comment

File: src/Common/Storage/Commands.Storage/Blob/Cmdlet/NewAzureStorageBlobSasToken.cs (hunk 2 above)
Comment (blueww) [inline src/Common/Storage/Commands.Storage/Blob/Cmdlet/NewAzureStorageBlobSasToken.cs]: SharedAccessProtocol is an Enum, and Powershell support Enum well. Use can just tab to get the enum value, don't need to create by themselves.

File: src/Common/Storage/Commands.Storage/Blob/Cmdlet/NewAzureStorageContainerSasToken.cs (hunk 3 above)
Comment (blueww) [inline src/Common/Storage/Commands.Storage/Blob/Cmdlet/NewAzureStorageContainerSasToken.cs]: SharedAccessProtocol is an Enum, and Powershell support Enum well. Use can just tab to get the enum value, don't need to create by themselves.

**Comment by markcowl** [comment]
@blueww I'm excited to see this feature.  It looks like a little bit of rethinking input parmater types is necessary to make the cmdlets usable - the user should not have to separately create a complex .Net type in order to use the cmdlet.

Also, need to see some test coverage and I assume the help update covers the new cmdlets.

File: src/Common/Storage/Commands.Storage/Common/Cmdlet/NewAzureStorageAccountSasToken.cs (hunk 4 above)
Comment (blueww) [inline src/Common/Storage/Commands.Storage/Common/Cmdlet/NewAzureStorageAccountSasToken.cs]: SharedAccessAccountServices , SharedAccessAccountResourceTypes , SharedAccessProtocol is an Enum, and Powershell support Enum well. Use can just tab to get the enum value, don't need to create by themselves.

File: src/Common/Storage/Commands.Storage/Common/Cmdlet/NewAzureStorageAccountSasToken.cs (hunk 5 above)
Comment (blueww) [inline src/Common/Storage/Commands.Storage/Common/Cmdlet/NewAzureStorageAccountSasToken.cs]: SharedAccessAccountServices , SharedAccessAccountResourceTypes , SharedAccessProtocol is an Enum, and Powershell support Enum well. Use can just tab to get the enum value, don't need to create by themselves.

File: src/Common/Storage/Commands.Storage/Common/Cmdlet/NewAzureStorageAccountSasToken.cs (hunk 6 above)
Comment (blueww) [inline src/Common/Storage/Commands.Storage/Common/Cmdlet/NewAzureStorageAccountSasToken.cs]: SharedAccessAccountServices , SharedAccessAccountResourceTypes , SharedAccessProtocol is an Enum, and Powershell support Enum well. Use can just tab to get the enum value, don't need to create by themselves.

File: src/Common/Storage/Commands.Storage/File/Cmdlet/NewAzureStorageFileSasToken.cs (hunk 8 above)
Comment (blueww) [inline src/Common/Storage/Commands.Storage/File/Cmdlet/NewAzureStorageFileSasToken.cs]: SharedAccessAccountServices , SharedAccessAccountResourceTypes , SharedAccessProtocol is an Enum, and Powershell support Enum well. Use can just tab to get the enum value, don't need to create by themselves.

File: src/Common/Storage/Commands.Storage/Common/Cmdlet/NewAzureStorageAccountSasToken.cs (hunk 7 above)
Comment (blueww) [inline src/Common/Storage/Commands.Storage/Common/Cmdlet/NewAzureStorageAccountSasToken.cs]: "ExpiryTime" is standard name in azure rest API, and consistent with other new SAS token cmdlets which already exist for a long time. So we don't suggest to change that.

File: src/Common/Storage/Commands.Storage/File/Cmdlet/NewAzureStorageShareSasToken.cs (hunk 9 above)
Comment (blueww) [inline src/Common/Storage/Commands.Storage/File/Cmdlet/NewAzureStorageShareSasToken.cs]: SharedAccessAccountServices , SharedAccessAccountResourceTypes , SharedAccessProtocol is an Enum, and Powershell support Enum well. Use can just tab to get the enum value, don't need to create by themselves.

File: src/Common/Storage/Commands.Storage/Queue/Cmdlet/NewAzureStorageQueueSasToken.cs (hunk 10 above)
Comment (blueww) [inline src/Common/Storage/Commands.Storage/Queue/Cmdlet/NewAzureStorageQueueSasToken.cs]: SharedAccessAccountServices , SharedAccessAccountResourceTypes , SharedAccessProtocol is an Enum, and Powershell support Enum well. Use can just tab to get the enum value, don't need to create by themselves.

File: src/Common/Storage/Commands.Storage/Table/Cmdlet/NewAzureStorageTableSasToken.cs (hunk 11 above)
Comment (blueww) [inline src/Common/Storage/Commands.Storage/Table/Cmdlet/NewAzureStorageTableSasToken.cs]: SharedAccessAccountServices , SharedAccessAccountResourceTypes , SharedAccessProtocol is an Enum, and Powershell support Enum well. Use can just tab to get the enum value, don't need to create by themselves.

File: src/Common/Storage/Commands.Storage/Blob/Cmdlet/NewAzureStorageContainerSasToken.cs (hunk 1 above)
Comment (blueww) [inline src/Common/Storage/Commands.Storage/Blob/Cmdlet/NewAzureStorageContainerSasToken.cs]: fixed

**Comment by blueww** [comment]
Hi Mark,

Thanks for the review and comments!

I have replied all your comments. Summarize as below:

1.	For the parameter types  “SharedAccessAccountServices” , “SharedAccessAccountResourceTypes” , “SharedAccessProtocol”, they are all Enum, and Powershell support Enum input very well. User can just press tab to get the enum value, don't need to create by themselves. Such as following parameters works well:
New-AzureStorageAccountSASToken -Service Blob,File,Table,Queue -ResourceType Container,Object,Service -Permission rw -Protocol HttpsOrHttp

2.	You have a comments to change “ExpiryTime” to “ExpirationTime”. As "ExpiryTime" is standard name in azure rest API, and release long ago in cmdlets to create SAS token, we don't plan to change that.
3.	I have fixed the indentation issue raised in your comments in the PR.
4.	We have already complete the test implementation of the new features, and 100% passed. See the attached report.
5.	Yes, the help file has already been changed to include the new support features.

Please let me know if you have other questions or comments!
Thanks!

File: src/ResourceManager/KeyVault/Commands.KeyVault/Commands.KeyVault.csproj [hunk 12]
```diff
@@ -183,6 +183,9 @@
       <SpecificVersion>False</SpecificVersion>
       <HintPath>..\..\..\packages\Newtonsoft.Json.6.0.8\lib\net45\Newtonsoft.Json.dll</HintPath>
     </Reference>
+    <Reference Include="System.Spatial, Version=5.6.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
```
Comment (hovsepm) [inline src/ResourceManager/KeyVault/Commands.KeyVault/Commands.KeyVault.csproj]: Why do you have changes here?

File: ChangeLog.md [hunk 13]
```diff
@@ -12,6 +12,16 @@
...
+ * Azure Storage
+  * Add cmdlet to generate SAS token against storage account
+    * New-AzureStorageAccountSASToken
+  * Add IPAddressOrRange/Protocol support in cmdlets to generate SAS token against blob, container, file, share, table, queue
+    * New-AzureStorageBlobSASToken
+    * New-AzureStorageContainerSASToken
+    * New-AzureStorageFileSASToken
+    * New-AzureStorageShareSASToken
+    * New-AzureStorageQueueSASToken
+    * New-AzureStorageTableSASToken
```
Comment (hovsepm) [inline ChangeLog.md]: align indentation.

**Pull request merged**